[core] add flare feature (contact support)

degemer opened the pull request:
- Create a Flare class to collect all relevant files (logs &
  configuration files), and then upload them to datadog, which will
  transfer them to support
- Put all code relative to the configcheck in a function for reusability
- add the flare option to init.d script

As I said, it adds some imports and probably belongs somewhere else. (at least to be easily accessible from the GUI for Windows/Mac)

LeoCavaille commented:
What about starting to have a real `utils/` subclass where we can migrate utils. I hate this toplevel `util` thingie.
`from utils.flare import Flare` ?

LeoCavaille commented on agent.py:
```diff
@@ -22,16 +22,26 @@
+from checks.check_status import CollectorStatus, DogstatsdStatus, ForwarderStatus
+from config import get_config, get_system_stats, get_parsed_args,\
+                   load_check_directory, get_logging_config, check_yaml,\
+                   get_config, get_config_path, get_confd_path
```
`get_config` duped. To avoid that even better:
```python
from config import (
     sorted symbols a-z
     ...
)

LeoCavaille commented on agent.py:
```diff
@@ -22,16 +22,26 @@
+import json
+import tempfile
+import re
+import atexit
```
:a: - :zzz: FTW

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+    def _ask_for_confirmation(self):
+        print '{0} is going to be uploaded to Datadog.'.format(self._tar_path)
+        print 'Do you want to continue [Y/n]?',
+        choice = raw_input().lower()
```
`choice = raw_input('Do you want to continue [Y/n]? ').lower()`

LeoCavaille commented:
Info that could be added:
* pip freeze from the self-contained agent directory ?
* version string from config.py (in case the pickle stuff doesn't work)

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+    # Save logs file paths
+    def _save_logs_path(self, config):
+        prefix = ''
+        if get_os() == 'windows':
```
Can you use `Platform.is_windows()` instead?

remh replied:
Also you should probably use the get_logging_config function https://github.com/DataDog/dd-agent/blob/master/config.py#L900

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+            os.path.join(self._prefix, 'etc', 'datadog.conf')
+        )
+
+        if get_os() != 'windows':
```
Same here.

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+
+    # Print supervisor status (and nothing on windows)
+    def _supervisor_status(self):
+        if get_os == 'windows':
```
Same here.

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+    # Start by creating the tar file which will contain everything
+    def _init_tarfile(self):
+        # Default temp path
+        self._tar_path = os.path.join(tempfile.gettempdir(), 'datadog-agent.tar.bz2')
```
Can we have an epoch in the filename instead of erasing the old one?

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+        if self._cmdline:
+            print output
+        else:
+            log.info(output)
```
Can't we just set up the logger to have a stdout handler in that case?

degemer replied:
Sure, but the ouput is not really beautiful.
> 2015-03-12 20:25:54,330 | INFO | dd.collector | collector(dd-agent:464) | Collecting logs and configuration files:
> 2015-03-12 20:25:54,331 | INFO | dd.collector | collector(dd-agent:464) |   * /var/log/datadog/collector.log.1

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+        elif resp.status_code == 500:
+            raise Exception('An error has occurred while uploading: {0}'.format(resp.text))
+        else:
+            raise Exception('An unknown error has occured, please email support directly')
```
Still output something in that case to help support debug flare?

remh commented on agent.py:
```diff
@@ -22,16 +22,26 @@
 import glob
+import tarfile
+import subprocess
+import json
```
You should use simplejson

remh commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+        self._save_logs_path(get_logging_config())
+        config = get_config()
+        self._api_key = config.get('api_key')
+        self._url = "{0}{1}".format(config.get('dd_url'), self.DATADOG_SUPPORT_URL)
```
Nitpick but you should use urljoin instead

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+
+        email = self._ask_for_email()
+
+        self._print("Uploading {0} to Datadog Support".format(self._tar_path))
```
Should we have some sort of limitation / output the size of the file? If the logs are really huge (like in debug mode with lots of errors) would this request succeed?

degemer replied:
The compression ratio is really good (only text files), on my tests 17MB -> 350KB, so it depends how huge it is. I could set an arbitrary limit on the size anyway.

LeoCavaille commented on agent.py:
```diff
@@ -196,6 +206,266 @@ def _do_restart(self):
+
+        for line in orig_file.splitlines(True):
+            if not self.COMMENT_REGEX.match(line) and not self.APIKEY_REGEX.match(line):
+                temp_file.write(line)
```
We can probably redact only part of the API key, `XXXXXXXXXXXXXXXXXXXXXXbcdef` so we can still have a good chance of finding it in the DB without compromising it?

LeoCavaille commented:
Awesome project @degemer! Looking forward to **using it** and **saving a lot of time** for our customers and ourselves. cc @DBLaw

remh commented:
+1 great job overall @degemer !

degemer commented:
Thanks @LeoCavaille  & @remh !
Léo: even if the pickle isn't there, the version string is shown. I will have a look tomorrow for `pip freeze`.
I think I addressed all your comments, here is an example output:
```
2015-03-12 22:51:06,106 | INFO | dd.collector | flare(flare.py:91) | Collecting logs and configuration files:
2015-03-12 22:51:06,107 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/collector.log
2015-03-12 22:51:06,111 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/forwarder.log.1
2015-03-12 22:51:07,767 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/forwarder.log
2015-03-12 22:51:08,330 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/dogstatsd.log
2015-03-12 22:51:08,637 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/dogstatsd.log.1
2015-03-12 22:51:08,650 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/jmxfetch.log
2015-03-12 22:51:08,876 | INFO | dd.collector | flare(flare.py:163) |   * /var/log/datadog/supervisord.log
2015-03-12 22:51:08,890 | INFO | dd.collector | flare(flare.py:172) |   * /etc/dd-agent/datadog.conf
2015-03-12 22:51:08,892 | INFO | dd.collector | flare(flare.py:183) |   * /etc/dd-agent/supervisor.conf
2015-03-12 22:51:08,907 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/fluentd.yaml
2015-03-12 22:51:08,912 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/jmx.yaml
2015-03-12 22:51:08,920 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/activemq.yaml - this file contains a password which has been removed in the version collected
2015-03-12 22:51:08,924 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/etcd.yaml
2015-03-12 22:51:08,927 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/mysql.yaml - this file contains a password which has been removed in the version collected
2015-03-12 22:51:08,934 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/consul.yaml
2015-03-12 22:51:08,936 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/nginx.yaml
2015-03-12 22:51:08,939 | INFO | dd.collector | flare(flare.py:211) |   * /etc/dd-agent/conf.d/network.yaml
2015-03-12 22:51:08,940 | INFO | dd.collector | flare(flare.py:95) |   * datadog-agent configcheck output
2015-03-12 22:51:09,017 | INFO | dd.collector | flare(flare.py:97) |   * datadog-agent status output
2015-03-12 22:51:10,102 | INFO | dd.collector | flare(flare.py:99) |   * datadog-agent info output
2015-03-12 22:51:10,286 | INFO | dd.collector | flare(flare.py:102) | Saving all files to /tmp/datadog-agent-2015-03-12-22-51-06.tar.bz2
/tmp/datadog-agent-2015-03-12-22-51-06.tar.bz2 is going to be uploaded to Datadog.
Do you want to continue [Y/n]?
Please enter your email: *****@datadoghq.com
2015-03-12 22:51:19,913 | INFO | dd.collector | flare(flare.py:114) | Uploading /tmp/datadog-agent-2015-03-12-22-51-06.tar.bz2 to Datadog Support
2015-03-12 22:51:20,360 | INFO | dd.collector | flare(flare.py:300) | Your logs were successfully uploaded. For future reference, your internal case id is 666
```

degemer commented:
I added the `pip freeze` output.

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,313 @@
+            'hostname': self._hostname,
+            'email': email
+        }
+        r = requests.post(url, files=files, data=data)
```
Could we pass the apikey as a query string parameter ?

degemer commented:
- Send `api_key` as a query parameter. (@remh comment)
- Update flare url to something more generic

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,313 @@
+
+    # Print output (success/error) of the request
+    def _analyse_result(self, resp):
+        if resp.status_code == 200:
```
Nitpick but you should likely use:
```python
resp.raise_for_status():
```
It will: 
- Make sure that if one day we change the response code from 200 to 202 (which we should), your code would still work
- Format an exception message in a nicer way so you won't have to do it yourself.

degemer replied:
Actually I try to send explicit errors (like 'Apikey unknown' , 'This case is not yours') from the server, which I then catch and display whereas `raise_for_status` only gives the status code and the standard error associated. But I should use ranges for status code.

remh commented:
Looks almost good! Could you add a test to make sure that the endpoint exist please?

degemer commented:
I added some test for the `upload`, and another on testing the existence of the endpoint (marked as failing for now).
First try with unittest, there is probably a better way to write them.

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+    # Print output (success/error) of the request
+    def _analyse_result(self):
+        try:
+            json_resp = json.loads(self._resp.text)
```
You coud just do:
```python
json_resp = self._resp.json()
```

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+        _, temp_path = tempfile.mkstemp(prefix='dd')
+        atexit.register(os.remove, temp_path)
+        temp_file = open(temp_path, 'w')
+        orig_file = open(file_path, 'r').read()
```
It looks like you are not closing orig_file.

You should use python "with" syntax to handle files properly

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+        temp_file = open(temp_path, 'w')
+        orig_file = open(file_path, 'r').read()
+        password_found = ''
+        for line in orig_file.splitlines(True):
```
You should use orig_file.readlines() instead

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+            print 'Windows - status not implemented'
+        else:
+            print '/etc/init.d/datadog-agent status'
+            self._print_output_command(['/etc/init.d/datadog-agent', 'status'])
```
This will fail on source install for example

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+            print '/etc/init.d/datadog-agent status'
+            self._print_output_command(['/etc/init.d/datadog-agent', 'status'])
+            print 'supervisorctl status'
+            self._print_output_command(['/opt/datadog-agent/bin/supervisorctl',
```
Same here

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+        except ValueError, e:
+            raise Exception('An unknown error has occured: {0}\n'\
+                            'Please contact support by email'.format(self._resp.text))
+        if self._resp.status_code in range(200, 203):
```
Why not using self._resp.raise_for_status() ?

remh commented on utils/flare.py:
```diff
@@ -0,0 +1,322 @@
+        temp_file = open(temp_path, 'w')
+        orig_file = open(file_path, 'r').read()
+
+        for line in orig_file.splitlines(True):
```
Same thing here, you should use the "with" syntax and readlines()

remh commented on packaging/debian/datadog-agent.init:
```diff
@@ -154,9 +154,15 @@ case "$1" in
         exit $?
         ;;
 
+    flare)
```
Looks like you didn't add it on the centos init file.

remh replied:
You should also add it in the source install script that manages the agent.

remh commented:
Awesome!

LeoCavaille commented:
Sweeeeet! :100: 
I am so looking forward to seeing it used on support cases!

Also protip @degemer : I did it for you but don't forget to delete your branches once they're merged to master.
